packages/flutter/lib/src/widgets/widget_state.dart

```diff
@@ -0,0 +1,653 @@
+///
+/// If a [WidgetStateTextStyle] is used for a property or a parameter that doesn't
+/// support resolving [WidgetStateProperty<TextStyle>]s, then its default color
+/// value will be used for all states.
+///
+/// To define a `const` [WidgetStateTextStyle], you'll need to extend
+/// [WidgetStateTextStyle] and override its [resolve] method. You'll also need
+/// to provide a `defaultValue` to the super constructor, so that we can know
+/// at compile-time what its default color is.
+/// See also:
+///
+///  * [MaterialStateTextStyle], the Material specific version of
+///    `WidgetStateTextStyle`.
+abstract class WidgetStateTextStyle extends TextStyle implements WidgetStateProperty<TextStyle> {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const WidgetStateTextStyle();
+
+  /// Creates a [WidgetStateTextStyle] from a [WidgetPropertyResolver<TextStyle>]
+  /// callback function.
+  ///
+  /// If used as a regular text style, the style resolved in the default state (the
+  /// empty set of states) will be used.
+  ///
+  /// The given callback parameter must return a non-null text style in the default
+  /// state.
+  static WidgetStateTextStyle resolveWith(WidgetPropertyResolver<TextStyle> callback) =>
+      _WidgetStateTextStyle(callback);
+
+  /// Returns a [TextStyle] that's to be used when a component is in the
+  /// specified state.
+  @override
+  TextStyle resolve(Set<WidgetState> states);
+}
+
+/// A [WidgetStateTextStyle] created from a [WidgetPropertyResolver<TextStyle>]
+/// callback alone.
+///
+/// If used as a regular text style, the style resolved in the default state will
+/// be used.
+///
+/// Used by [WidgetStateTextStyle.resolveWith].
+class _WidgetStateTextStyle extends WidgetStateTextStyle {
+  const _WidgetStateTextStyle(this._resolve);
+
+  final WidgetPropertyResolver<TextStyle> _resolve;
+
+  @override
+  TextStyle resolve(Set<WidgetState> states) => _resolve(states);
+}
+
+/// Interface for classes that [resolve] to a value of type `T` based
+/// on a widget's interactive "state", which is defined as a set
+/// of [WidgetState]s.
+///
+/// Widget state properties represent values that depend on a widget's "state".
+/// The state is encoded as a set of [WidgetState] values, like
+/// [WidgetState.focused], [WidgetState.hovered], [WidgetState.pressed]. For
+/// example the [InkWell.overlayColor] defines the color that fills the ink well
+/// when it's pressed (the "splash color"), focused, or hovered. The [InkWell]
+/// uses the overlay color's [resolve] method to compute the color for the
+/// ink well's current state.
+///
+/// [ButtonStyle], which is used to configure the appearance of
+/// buttons like [TextButton], [ElevatedButton], and [OutlinedButton],
+/// has many material state properties. The button widgets keep track
+/// of their current material state and [resolve] the button style's
```

Comment:
```suggestion
/// of its current material state and [resolve] the button style's
```

Reply:
Shouldn't it be "their" because of "button widgets"? (I am not exactly the best with grammar)